pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,437 @@
+// nil is returned.
+func addressToIngress(dnsName *string, ipAddress *string) *corev1.LoadBalancerIngress {
+	var ingress *corev1.LoadBalancerIngress
+
+	if ipAddress != nil || dnsName != nil {
+		ingress = &corev1.LoadBalancerIngress{}
+		if dnsName != nil {
+			ingress.Hostname = *dnsName
+		}
+
+		if ipAddress != nil {
+			ingress.IP = *ipAddress
+		}
+	}
+
+	return ingress
+}
+
+func ensureAssociateFIPWithInstance(openstackClientFactory openstackclient.Factory, instance *servers.Server, floatingIP *floatingips.FloatingIP) error {
+	fipid, err := findFloatingIDByInstanceID(openstackClientFactory, instance.ID)
+	if err != nil {
+		return err
+	}
+
+	if fipid != "" {
+		return nil
+	}
+
+	if floatingIP.Status != "ACTIVE" || instance.Status != "ACTIVE" {
+		return fmt.Errorf("instance or public ip address not ready yet")
+	}
+
+	associateOpts := computefip.AssociateOpts{
+		FloatingIP: floatingIP.FloatingIP,
+	}
+
+	err = associateFIPWithInstance(openstackClientFactory, instance.ID, associateOpts)
+	if err != nil {
+		return fmt.Errorf("failed to associate public ip address %s to instance %s: %w", floatingIP.FloatingIP, instance.Name, err)
+	}
+	return nil
+}
+
+func ensureSecurityGroupRules(openstackClientFactory openstackclient.Factory, bastion *extensionsv1alpha1.Bastion, opt *Options, secGroupID string) error {
+	ingressPermissions, err := ingressPermissions(bastion)
+	if err != nil {
+		return err
+	}
+
+	shootSecurityGroups, err := getSecurityGroups(openstackClientFactory, opt.ShootName)
+	if err != nil {
+		return err
+	}
+
+	if len(shootSecurityGroups) == 0 {
+		return errors.New("shootSecurityGroups must not be empty")
+	}
+	// The assumption is that the shoot only has one security group
+	shootSecurityGroupID := shootSecurityGroups[0].ID
+
+	var wantedRules []rules.CreateOpts
+	for _, ingressPermission := range ingressPermissions {
+		wantedRules = append(wantedRules,
+			IngressAllowSSH(opt, ingressPermission.EtherType, secGroupID, ingressPermission.CIDR),
+			EgressAllowSSHToWorker(opt, secGroupID, shootSecurityGroupID),
+		)
+	}
+
+	currentRules, err := listRules(openstackClientFactory, secGroupID)
+	if err != nil {
+		return fmt.Errorf("failed to list rules: %w", err)
+	}
+
+	rulesToAdd, rulesToDelete := rulesSymmetricDifference(wantedRules, currentRules)
+
+	for _, rule := range rulesToAdd {
+		if err := createSecurityGroupRuleIfNotExist(openstackClientFactory, rule); err != nil {
+			return fmt.Errorf("failed to add security group rule %s: %w", rule.Description, err)
+		}
+	}
+
+	for _, rule := range rulesToDelete {
+		err := deleteRule(openstackClientFactory, rule.ID)
+
+		if err != nil {
+			if !openstackclient.IsNotFoundError(err) {
+				return fmt.Errorf("failed to delete security group rule %s (%s): %w", rule.Description, rule.ID, err)
+			} // else: already deleted
+		} else {
+			logger.Info("Unwanted security group rule deleted", "rule", rule.Description, "ruleID", rule.ID)
+		}
```

Comment:
```suggestion
		if err := deleteRule(openstackClientFactory, rule.ID); err != nil {
			if openstackclient.IsNotFoundError(err) {
				continue
			}
			return fmt.Errorf("failed to delete security group rule %s (%s): %w", rule.Description, rule.ID, err)
		}
		logger.Info("Unwanted security group rule deleted", "rule", rule.Description, "ruleID", rule.ID)
```
nit

Reply:
thanks, committed in ebec7c7f per comments